File: ts/lang_chat/agent/agent.ts
/**
 * LangChat Tutor Agent — TypeScript
 *
 * This file defines the tutor agent using LangChain's createAgent.
 * It is the main entry point for LangGraph Cloud deployment.
 *
 * The agent uses:
 * - Gemini (default) via initChatModel — swap to any supported provider by uncommenting below
 * - readLessonMaterial tool — reads lesson content from the tutor_lx/ directory
 * - LangChain MCP docs server — for documentation lookups beyond the lesson material
 *
 * Assistant context (set per assistant in LangSmith):
 * - lessonId:       which lesson to load instructions and material for (e.g. "tutor_l1")
 * - storeNamespace: student namespace (first_last), used to look up store data
 * - studentName:    student's first name, used to personalize responses
 */

import path from "node:path";
import fs from "node:fs";
import { fileURLToPath } from "node:url";
import * as z from "zod";

import { createAgent, dynamicSystemPromptMiddleware, tool, type ToolRuntime } from "langchain";
import { initChatModel } from "langchain/chat_models/universal";
import { MultiServerMCPClient } from "@langchain/mcp-adapters";

import { SYSTEM_PROMPT as BASE_SYSTEM_PROMPT } from "./systemPrompt.js";

// ---------------------------------------------------------------------------
// Model — uncomment the provider you want to use
// All provider packages are installed; set the matching API key in .env
// ---------------------------------------------------------------------------
const model = await initChatModel("google_genai:gemini-2.5-flash");    // Gemini (default — generous free tier)
// const model = await initChatModel("openai:gpt-4o");                 // OpenAI
// const model = await initChatModel("anthropic:claude-sonnet-4-6");   // Anthropic
// const model = await initChatModel("ollama:llama3.2");               // Ollama (local)

const LESSONS_DIR = path.join(path.dirname(fileURLToPath(import.meta.url)), "..");


// ---------------------------------------------------------------------------
// Context schema — injected by LangGraph Cloud from assistant context dict
// ---------------------------------------------------------------------------
const contextSchema = z.object({
  lessonId: z.string().default("tutor_l1"),
  studentName: z.string().default("Student"),
  storeNamespace: z.string().default(""),
});

type Context = z.infer<typeof contextSchema>;


function loadLessonInstructions(lessonId: string): string {
  const instructionsFile = path.join(LESSONS_DIR, lessonId, `${lessonId}_instructions.md`);
  if (!fs.existsSync(instructionsFile)) {
    return `(No instructions file found for lesson '${lessonId}')`;
  }
  return fs.readFileSync(instructionsFile, "utf-8");
}


// ---------------------------------------------------------------------------
// Dynamic system prompt — reads lessonId and studentName from assistant context
// ---------------------------------------------------------------------------
const lessonPrompt = dynamicSystemPromptMiddleware<Context>((state, runtime) => {
  const lessonId = runtime.context.lessonId;
  const studentName = runtime.context.studentName;
  const instructions = loadLessonInstructions(lessonId);
  return (
    `${BASE_SYSTEM_PROMPT}\n\n` +
    `## Current session\n` +
    `Student: ${studentName}\n` +
    `Lesson: ${lessonId}\n\n` +
    `## Lesson instructions\n\n` +
    instructions
  );
});


// ---------------------------------------------------------------------------
// Lesson material tool
// Reads the information file for the current lesson from assistant context
// ---------------------------------------------------------------------------
const readLessonMaterial = tool(
  async (_: object, runtime: ToolRuntime<Context>) => {
    const lessonId = runtime.context.lessonId;
    const infoFile = path.join(LESSONS_DIR, lessonId, `${lessonId}_information.md`);
    if (!fs.existsSync(infoFile)) {
      return `Lesson material not found for '${lessonId}'.`;
    }
    return fs.readFileSync(infoFile, "utf-8");
  },
  {
    name: "read_lesson_material",
    description:
      "Read the reference material for the current lesson. " +
      "Use this tool when you need to answer a student's question or quiz them. " +
      "Always check the lesson material before answering.",
    schema: z.object({}),
  }
);


// ---------------------------------------------------------------------------
// MCP tools — LangChain/LangGraph/LangSmith docs search
// ---------------------------------------------------------------------------
const mcpClient = new MultiServerMCPClient({
  "docs": {
    url: "https://docs.langchain.com/mcp",
    transport: "streamable_http",
  },
});

const mcpTools = await mcpClient.getTools();


// ---------------------------------------------------------------------------
// Agent
// ---------------------------------------------------------------------------
export const graph = createAgent({
  model,
  tools: [readLessonMaterial, ...mcpTools],
  middleware: [lessonPrompt],
  contextSchema,
});

File: python/lang_chat/agent/agent.py
"""
LangChat Tutor Agent — Python

This file defines the tutor agent using LangChain's create_agent.
It is the main entry point for LangGraph Cloud deployment.

The agent uses:
- Gemini (default) via init_chat_model — swap to any supported provider by uncommenting below
- read_lesson_material tool — reads lesson content from the tutor_lx/ directory

Assistant context (set per assistant in LangSmith):
- lesson_id:       which lesson to load instructions and material for (e.g. "tutor_l1")
- store_namespace: student namespace (first_last), used to look up store data
- student_name:    student's first name, used to personalize responses
"""

import asyncio
from dataclasses import dataclass
from pathlib import Path

from langchain.agents import create_agent
from langchain.agents.middleware import ModelRequest, dynamic_prompt
from langchain.chat_models import init_chat_model
from langchain.tools import ToolRuntime, tool
from langchain_mcp_adapters.client import MultiServerMCPClient

import sys
sys.path.insert(0, str(Path(__file__).parent))
from system_prompt import BASE_SYSTEM_PROMPT

# ---------------------------------------------------------------------------
# Model — uncomment the provider you want to use
# All provider packages are installed; set the matching API key in .env
# ---------------------------------------------------------------------------
model = init_chat_model("google_genai:gemini-2.5-flash")        # Gemini (default)
# model = init_chat_model("openai:gpt-4o")                    # OpenAI
# model = init_chat_model("anthropic:claude-sonnet-4-6")      # Anthropic
# model = init_chat_model("ollama:llama3.2")                  # Ollama (local)

LESSONS_DIR = Path(__file__).parent.parent


# ---------------------------------------------------------------------------
# Context schema — injected by LangGraph Cloud from assistant context dict
# ---------------------------------------------------------------------------
@dataclass
class ContextSchema:
    lesson_id: str = "tutor_l1"
    student_name: str = "Student"
    store_namespace: str = ""


def _load_lesson_instructions(lesson_id: str) -> str:
    instructions_file = LESSONS_DIR / lesson_id / f"{lesson_id}_instructions.md"
    if not instructions_file.exists():
        return f"(No instructions file found for lesson '{lesson_id}')"
    return instructions_file.read_text()


# ---------------------------------------------------------------------------
# Dynamic system prompt — reads lesson_id and student_name from assistant context
# ---------------------------------------------------------------------------
@dynamic_prompt
def lesson_prompt(request: ModelRequest) -> str:
    lesson_id = request.runtime.context.lesson_id
    student_name = request.runtime.context.student_name
    instructions = _load_lesson_instructions(lesson_id)
    return (
        f"{BASE_SYSTEM_PROMPT}\n\n"
        f"## Current session\n"
        f"Student: {student_name}\n"
        f"Lesson: {lesson_id}\n\n"
        f"## Lesson instructions\n\n"
        f"{instructions}"
    )


# ---------------------------------------------------------------------------
# Lesson material tool
# Reads the information file for the current lesson from assistant context
# ---------------------------------------------------------------------------
@tool
def read_lesson_material(runtime: ToolRuntime[ContextSchema]) -> str:
    """Read the reference material for the current lesson.

    Use this tool when you need to answer a student's question or quiz them.
    Always check the lesson material before answering.
    """
    lesson_id = runtime.context.lesson_id
    info_file = LESSONS_DIR / lesson_id / f"{lesson_id}_information.md"
    if not info_file.exists():
        return f"Lesson material not found for '{lesson_id}'."
    return info_file.read_text()


# ---------------------------------------------------------------------------
# MCP tools — LangChain/LangGraph/LangSmith docs search
# ---------------------------------------------------------------------------
_mcp_client = MultiServerMCPClient({
    "docs": {
        "url": "https://docs.langchain.com/mcp",
        "transport": "http",
    }
})
mcp_tools = asyncio.run(_mcp_client.get_tools())


# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------
graph = create_agent(
    model=model,
    tools=[read_lesson_material, *mcp_tools],
    middleware=[lesson_prompt],
    context_schema=ContextSchema,
)
